File: Code Examples/L06-07. Problems, Algorithms and Programs/Structured Language. C/anagrams.c

#include <stdio.h>
#include <stdlib.h>

//--------------------------------------------------
// initialise_array
//--------------------------------------------------
int* initialise_array(int num, int n) {
	int* res = (int*)malloc(n * sizeof(int));

	for (int i = (n - 1); i >= 0; i--) {
		res[i] = num % 10;
		num = num / 10;
	}

	return res;
}

//--------------------------------------------------
// bubble_sort
//--------------------------------------------------
void bubble_sort(int* l, int n) {
	for (int i = 0; i < (n - 1); i++)
		for (int j = 0; j < ((n - 1) - i); j++)
			if (l[j] > l[j + 1]) {
				int temp = l[j];
				l[j] = l[j + 1];
				l[j + 1] = temp;
			}
}

//--------------------------------------------------
// are_equal
//--------------------------------------------------
int are_equal(int* l1, int* l2, int n) {
	int res = 1; 
	int i = 0; 

	while ((i < n) && (res == 1)) {
		if (l1[i] != l2[i])
			res = 0;
		else
			i = i + 1;
	}

	return res;
}

//--------------------------------------------------
// anagrams2
//--------------------------------------------------
int anagrams2(int* l1, int* l2, int n) {
	//Sort the numbers
	bubble_sort(l1, n);
	bubble_sort(l2, n);

	//Compare the result
	int res = are_equal(l1, l2, n);

	return res;
}

//--------------------------------------------------
// main
//--------------------------------------------------
int main(int argc, char* argv[]) {
	//Get the input numbers to arrays
	int num1 = atoi(argv[1]);
	int num2 = atoi(argv[2]);
	int n = atoi(argv[3]);

	int* l1 = initialise_array(num1, n);
	int* l2 = initialise_array(num2, n);

	//Call to anagrams2
	int res = anagrams2(l1, l2, n);

	//Free the memory 
	free(l1);
	free(l2);

	//Show the result to the user
	if (res == 1)
		printf("True\n");
	else
		printf("False\n");

	return 0;
}

File: Code Examples/L06-07. Problems, Algorithms and Programs/General Purpose Scripting Language. Python/anagrams.py

#--------------------------------------------------------
#           PYTHON PROGRAM
# Here is where we are going to define our set of...
# - Imports
# - Global Variables
# - Functions
# ...to achieve the functionality required.
# When executing > python 'this_file'.py in a terminal,
# the Python interpreter will load our program,
# but it will execute nothing yet.
#--------------------------------------------------------

import sys

#--------------------------------------------------
# initialise_array
#--------------------------------------------------
def initialise_array(num):
    l = []
    while num > 0:
        l.append(num % 10)
        num = num / 10

    i = 0
    j = len(l)-1
    while i < j:
        temp = l[i]
        l[i] = l[j]
        l[j] = temp
        i = i + 1
        j = j - 1

    return l

#--------------------------------------------------
# bubble_sort
#--------------------------------------------------
def bubble_sort(l):
    for i in range(0, len(l)-1):
        for j in range(0, len(l)-1-i):
            if l[j] > l[j + 1]:
                temp = l[j]
                l[j] = l[j + 1]
                l[j + 1] = temp

#--------------------------------------------------
# comp
#--------------------------------------------------
def are_equal(l1, l2):
    res = True
    i = 0
    while i < len(l1) and res == True:
        if l1[i] != l2[i]:
            res = False
        else:
            i = i+1

    return res

#--------------------------------------------------
# anagrams2
#--------------------------------------------------
def anagrams2(l1, l2):
    bubble_sort(l1)
    bubble_sort(l2)

    res = are_equal(l1, l2)

    return res

#---------------------------------------------------------------
#           PYTHON EXECUTION
# This is the main entry point to the execution of our program.
# It provides a call to the 'main function' defined in our
# Python program, making the Python interpreter to trigger
# its execution.
#---------------------------------------------------------------
if __name__ == '__main__':
    #Get the input numbers to arrays
    num1 = [1,2,3,6]
    num2 = [5,3,2,1]

    #Call to alg_n2
    r = anagrams2(num1, num2)

    #Show the result to the user
    print(r)

#anagrams.py 34765 76345
#anagrams.py 55667 66885

